RABBIT
======

RABBIT was developped for Reconstrucing Ancestral Blocks BIT by bit in multiparental populations, which has been extended for genotype imputaiton and genetic map construction. Tested in Matheamtica 11.0. 


##Getting started

1. Clone or download.

2. Create a new mathematica notebook in a work folder, and set path to the folder "RABBIT_Packages". 

3. ```Needs["MagicReconstruct`"]``` for haplotype reconstruciton. 

   ```Needs["MagicImpute`"]``` for genotype imputation

   ```Needs["MagicMap`"]``` and ```myParallelNeeds["MagicMap`"]``` for genetic map construction

4. Manual in the folder "RABBIT_Manual", examples in the folder "RABBIT_Example Scripts", and tutorials in the folder "RABBIT_Tutorial" 

##

##References

Zheng, C., Boer, M. P., and Eeuwijk, F. A. 2014. A general modeling framework for genome ancestral origins in multi-parental populations. Genetics. 198: 87-101.

Zheng, C. 2015. Modeling X-linked ancestral origins in multiparental populations. G3: Genes | Genomes | Genetics. 5: 777-801. 

Zheng, C., Boer, M. P., and Eeuwijk, F. A. 2015. Reconstruction of genome ancestry blocks in multiparental populations. Genetics. 200: 1073-1087. 

Zheng, C., Boer, M. P., and Eeuwijk, F. A. 2018. Recursive algorithms for modeling genome blocks in a fixed pedigree. G3: Genes | Genomes | Genetics. 8: 3231–3245

Zheng, C., Boer, M. P., and Eeuwijk, F. A. 2018. Accurate genotype imputation in multiparental populations from low-coverage sequence. Genetics. 210: 71-82.

Zheng, C., Boer, M. P., and Eeuwijk, F. A. 2019. Construction of genetic linkage maps in multiparental populations. Submitted. 

##
